test/test_resampler.cpp

```diff
@@ -1160,5 +1169,351 @@ TEST(cubeb, individual_methods)
   ASSERT_EQ(frames_needed2, 0u);
 }
 
+struct sine_wave_state {
+  float phase = 0.0f;
+  float frequency;
+  int sample_rate;
+  sine_wave_state(float freq, int rate) : frequency(freq), sample_rate(rate) {}
+};
+
+long
+data_cb(cubeb_stream * stream, void * user_ptr, void const * input_buffer,
+        void * output_buffer, long nframes)
+{
+  sine_wave_state * state = static_cast<sine_wave_state *>(user_ptr);
+  float * out = static_cast<float *>(output_buffer);
+  float phase_increment = 2.0f * M_PI * state->frequency / state->sample_rate;
+
+  for (int i = 0; i < nframes; i++) {
+    float sample = sin(state->phase);
+    state->phase += phase_increment;
+    if (state->phase > 2.0f * M_PI) {
+      state->phase -= 2.0f * M_PI;
+    }
+    out[i] = sample * 0.8;
+  }
+  return nframes;
+}
+
+// This implements 4.6.2 from "Standard for Digitizing Waveform Recorders"
+// (in particular Annex A), then returns the estimated amplitude, phase, and the
+// sum of squared error relative to a sine wave sampled at `sample_rate` and of
+// frequency `frequency`. This is also described in "Numerical methods for
+// engineers" chapter 19.1, and explained at
+// https://www.youtube.com/watch?v=afQszl_OwKo and videos of the same series.
+// In practice here we're sending a perfect 1khz sine wave into a good
+// resampler, and despite the resampling ratio being quite extreme sometimes,
+// we're expecting a very good fit.
+float
+fit_sine(const std::vector<float> & signal, float sample_rate, float frequency,
+         float & out_amplitude, float & out_phase)
+{
+  const size_t len = signal.size();
+  float phase_incr = 2.0 * M_PI * frequency / sample_rate;
```

Comment:
Rounding errors are multiplied by the sample count.
```suggestion
  double phase_incr = 2.0 * M_PI * frequency / sample_rate;
```